cmd/keymasterd/adminHandlers.go

```diff
@@ -0,0 +1,187 @@
+	w http.ResponseWriter, r *http.Request) {
+	if state.sendFailureToClientIfNonAdmin(w, r) {
+		return
+	}
+	//TODO: check method, must be POST
+	if r.Method != "POST" {
+		state.writeFailureResponse(w, r, http.StatusMethodNotAllowed, "")
+		return
+	}
+
+	err := r.ParseForm()
+	if err != nil {
+		logger.Printf("error parsing err=%s", err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		return
+	}
+	usersToAdd, ok := r.Form["username"]
+	if !ok {
+		state.writeFailureResponse(w, r, http.StatusBadRequest, "Required Parameters missing")
+		return
+	}
+	//It is a good idea to match?
+	for _, username := range usersToAdd {
+		matched, err := regexp.Match(`^[A-Za-z0-9-_.]+$`, []byte(username))
+		if err != nil {
+			logger.Printf("error parsing err=%s", err)
+			state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+			return
+		}
+		if !matched {
+			state.writeFailureResponse(w, r, http.StatusBadRequest, "Invalid Usernames found")
+			return
+		}
+	}
+	var newProfiles map[string]*userProfile
+	newProfiles = make(map[string]*userProfile)
+	//check if usernames already exist
+	for _, username := range usersToAdd {
+		profile, existing, fromCache, err := state.LoadUserProfile(username)
+		if err != nil {
+			logger.Printf("error parsing err=%s", err)
+			state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+			return
+		}
+		if existing {
+			state.writeFailureResponse(w, r, http.StatusBadRequest, "User exist in DB")
+			return
+		}
+		if fromCache {
+			state.writeFailureResponse(w, r, http.StatusServiceUnavailable, "Working in db disconnected mode, try again later")
+			return
+		}
+		newProfiles[username] = profile
+	}
+	for username, userProfile := range newProfiles {
+		err := state.SaveUserProfile(username, userProfile)
+		if err != nil {
+			logger.Printf("error Savinf Profile  err=%s", err)
+			state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+			return
+		}
+	}
+	//if html then redirect to users page, else return jsonOK
+	preferredAcceptType := getPreferredAcceptType(r)
+	switch preferredAcceptType {
+	case "text/html":
+
+		http.Redirect(w, r, usersPath, http.StatusFound)
+	default:
+		w.WriteHeader(200)
+		fmt.Fprintf(w, "OK\n")
+	}
+}
+
+func (state *RuntimeState) deleteUsersHandler(
+	w http.ResponseWriter, r *http.Request) {
+	if state.sendFailureToClientIfNonAdmin(w, r) {
+		return
+	}
+	//TODO: check method, must be POST
+	if r.Method != "POST" {
+		state.writeFailureResponse(w, r, http.StatusMethodNotAllowed, "")
+		return
+	}
+
+	err := r.ParseForm()
+	if err != nil {
+		logger.Printf("error parsing err=%s", err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		return
+	}
+	usersToDelete, ok := r.Form["username"]
+	if !ok {
+		state.writeFailureResponse(w, r, http.StatusBadRequest, "Required Parameters missing")
+		return
+	}
+	for _, username := range usersToDelete {
+		err = state.DeleteUserProfile(username)
```

Comment:
I think it would be prudent to sanitise this input as well.
In fact, why not move all the code from the start of the function until the point where you process the list of usernames into a utility function, and then have the add and delete handlers call the utility function?

I'm also inclined to change the handlers to just a single username, which will then make the utility function a good fit for adding the bootstrap OTP. I think bulk operations can be left to a simple script.